exchanges/subscription/template.go

```diff
@@ -63,86 +68,121 @@ func (l List) ExpandTemplates(e iExchange) (List, error) {
+		} else {
+			subs = append(subs, expanded...)
 		}
+	}
 
-		subCtx := &tplCtx{
-			S:              s,
-			AssetPairs:     ap,
-			PairSeparator:  recordSeparator,
-			AssetSeparator: groupSeparator,
-		}
+	return subs, err
+}
 
-		t, err := e.GetSubscriptionTemplate(s)
-		if err != nil {
-			return nil, err
-		}
-		if t == nil {
-			return nil, errInvalidTemplate
+func expandTemplate(e iExchange, s *Subscription, ap assetPairs, assets asset.Items) (List, error) {
+	if s.QualifiedChannel != "" {
+		return List{s}, nil
+	}
+
+	t, err := e.GetSubscriptionTemplate(s)
+	if err != nil {
+		return nil, err
+	}
+	if t == nil {
+		return nil, errInvalidTemplate
+	}
+
+	subCtx := &tplCtx{
+		S:              s,
+		PairSeparator:  recordSeparator,
+		AssetSeparator: groupSeparator,
+		BatchSize:      deviceControl + "BS",
+	}
+
+	switch s.Asset {
+	case asset.All:
+		subCtx.AssetPairs = ap
+	default:
+		// This deliberately includes asset.Empty to harmonise handling
+		subCtx.AssetPairs = assetPairs{
+			s.Asset: ap[s.Asset],
 		}
+		assets = asset.Items{s.Asset}
+	}
 
-		buf := &bytes.Buffer{}
-		if err := t.Execute(buf, subCtx); err != nil {
-			return nil, err
+	if len(s.Pairs) != 0 {
+		for a, pairs := range subCtx.AssetPairs {
+			if err := pairs.ContainsAll(s.Pairs, true); err != nil { //nolint:govet // Shadow, or gocritic will complain sloppyReassign
+				return nil, err
+			}
+			subCtx.AssetPairs[a] = s.Pairs
 		}
+	}
 
-		out := buf.String()
+	buf := &bytes.Buffer{}
+	if err := t.Execute(buf, subCtx); err != nil { //nolint:govet // Shadow, or gocritic will complain sloppyReassign
+		return nil, err
+	}
 
-		subAssets := assets
-		xpandPairs := strings.Contains(out, subCtx.PairSeparator)
-		if xpandAssets := strings.Contains(out, subCtx.AssetSeparator); xpandAssets {
-			if s.Asset != asset.All {
-				return nil, errAssetTemplateWithoutAll
-			}
-		} else {
-			if xpandPairs && (s.Asset == asset.All || s.Asset == asset.Empty) {
-				// We don't currently support expanding Pairs without expanding Assets for All or Empty assets, but we could; waiting for a use-case
-				return nil, errInvalidAssetExpandPairs
+	out := strings.TrimSpace(buf.String())
+
+	// Remove a single trailing AssetSeparator; don't use a cutset to avoid removing 2 or more
+	out = strings.TrimSpace(strings.TrimSuffix(out, subCtx.AssetSeparator))
+
+	xpandPairs := strings.Contains(out, subCtx.PairSeparator)
+
+	assetRecords := strings.Split(out, subCtx.AssetSeparator)
+	if len(assetRecords) != len(assets) {
+		return nil, fmt.Errorf("%w: Got %d; Expected %d", errAssetRecords, len(assetRecords), len(assets))
+	}
+
+	subs := List{}
+	for i, assetChannels := range assetRecords {
+		a := assets[i]
+		pairs := subCtx.AssetPairs[a]
+
+		batchSize := len(pairs) // Default to all pairs in one batch
+		if b := strings.Split(assetChannels, subCtx.BatchSize); len(b) > 2 {
+			return nil, fmt.Errorf("%w for %s", errTooManyBatchSize, a)
```

Comment:
Would you mind either pm or here, expanding on what's happening? For the test scenario that hits here, there's an intentional addition of more batches, but then I don't understand the _why_ of the `Default to all pairs in one batch` and this part
You're using a hardcoded len of 2 here and below, why? Why not store that in a var with a more descriptive term? In the below section, you change if there is a batch indicator, but you only want one?

Reply:
I think perhaps the cleanest explanation might belong in the docs for subscriptions:
Fixed 1e271b4b

Let me explain here, and then we can work out together how/where/what to document for others.
* BatchSize is a directive, which is optional.
* If the template doesn't use a batch size, then there is only one batch, which includes all pairs.
Hence line 147.
* We expect only one batchSize directive per assetRecord. So when we split on it, we have three possible situations:
 * We got more than 2 elements, meaning more than one directive, which is an error `too many BatchSize directives`
 * We got exactly 2 elements. The first part is the asset record, and the second part is the batch size. So we assign them like that
 * Or if we get only 1 element from the split, it means it doesn't include a batchSize. So the batch size remains the length of all the pairs. EXCEPT if the pairs are going to be expanded, and we're not batching, then each batch is one pair. The opposite of the default on line 147. So the two cases to consider to understand this EXCEPT are a subscription for All assets, which either wants one subscription for all pairs, and so batch size is len(pairs), OR a subscriptins that wants one subscription for each pair, and so batch size is 1.

Does that help explain the logic here?
If so, was there a specific "Ah Ha" moment in that verbose description I should make sure is in the comments next to the code?

Reply:
Thanks for writing that out! These are definitely the underlying rules I was seeking. 
Trying to think it through:
- I think `errTooManyBatchSize` turns to `errOneBatchSizePerAssetPair` to make the rule more solid
- The line `return nil, fmt.Errorf("%w for %s: Got %d; Expected %d", errPairRecords, a, len(pairLines), len(batches)) adds something like `pair line length must match batchSize number` again to highlight the rules of template generation

I think that addresses my confusions... Hopefully 😄 If batchSize is present, there can only be one per assetPair in the template. That if there is an batchSize, the count of the slices of pairLines must match the number of batches.

Reply:
Fixed edde8d01e

I've added an overall description at the top instead of trying to inline the comments as we go.
I think that this section lends itself better to a contextual block than code+comment style.

On item:
2. The issues that cause this can be very varied, so I don't want to suggest that this is about batchSize, especially when it isn't used most of the time. This code reads "batchSize centric", but that's really only because if there might be one, it's easier to normalise all paths to that. On that basis I've added a comment, but left the error alone.

Reply:
I'm happy with that, thanks!